File: February2022/Java/WordLadder.java

// @saorav21994
// TC : O(beginWord.length() * beginWord.length() * wordList.size())
// SC : O(n)

// Hard - Map + BFS Queue (explained by CodingDecoded July last year)

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Queue<String> queue = new LinkedList<>();
        HashMap<String, Integer> map = new HashMap<>();
        queue.offer(beginWord);
        map.put(beginWord, 1);
        while (queue.isEmpty() != true) {
            String word = queue.poll();
            int trans = map.get(word);
            for (String wordi : wordList) {
                if (map.containsKey(wordi) == false && diffByChar(wordi, word) == 1) {
                    if (wordi.equals(endWord) == true) {
                        return trans+1;
                    }
                    map.put(wordi, trans+1);
                    queue.offer(wordi);
                }
            }
        }
        return 0;
    }
    
    private int diffByChar(String a, String b) {
        int l = a.length();
        int count = 0;
        for (int i = 0; i < l; i++) {
            if (a.charAt(i) != b.charAt(i)) {
                count += 1;
                if (count == 2) return count;
            }
        }
        return count;
    }
}

File: February2022/Kotlin/WordLadder.kt
// Author : Shobhit Behl(shobhitbruh)
class Solution {
    fun ladderLength(beginWord: String, endWord: String, wordList: MutableList<String>): Int {
        val hm = HashMap<String, MutableList<String>>()
        val visited = HashMap<String, Boolean>()
        var n = false
        for (i in wordList.indices) {
            if (wordList[i] == beginWord) {
                n = true
            }
            hm[wordList[i]] = ArrayList()
            visited[wordList[i]] = false
        }
        if (!n) {
            wordList.add(beginWord)
            hm[beginWord] = ArrayList()
            visited[beginWord] = false
        }
        if (!hm.containsKey(endWord)) {
            return 0
        }
        for (i in wordList.indices) {
            for (j in i + 1 until wordList.size) {
                if (wordList[i].length == wordList[j].length) {
                    var p = 0
                    for (k in 0 until wordList[i].length) {
                        if (wordList[i][k] != wordList[j][k]) {
                            p++
                        }
                    }
                    if (p == 1) {
                        hm[wordList[i]]!!.add(wordList[j])
                        hm[wordList[j]]!!.add(wordList[i])
                    }
                }
            }
        }
        // System.out.println(hm);
        val li = LinkedList<String>()
        li.addLast(beginWord)
        var p = 0
        while (li.size > 0) {
            var size = li.size
            p++
            while (size-- > 0) {
                val rem = li.removeFirst()
                visited[rem] = true
                if (rem == endWord) {
                    return p
                }
                for (i in hm[rem]!!.indices) {
                    if (!visited[hm[rem]!![i]]!!) {
                        li.addLast(hm[rem]!![i])
                    }
                }
            }
        }
        return 0
    }
}
